cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -27,44 +34,81 @@ import (
+	errInsecureClientNoRequest        = fmt.Errorf("doRequest did not make a request using the withInsecureHTTPClient client")
+	errInvalidPEMData                 = fmt.Errorf("invalid PEM data")
+	errTrailingPEMData                = fmt.Errorf("trailing data after first PEM block")
+)
+
 // RetrieveValidatedClusterInfo connects to the API Server and tries to fetch the cluster-info ConfigMap
 // It then makes sure it can trust the API Server by looking at the JWS-signed tokens
-func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers []string) (*clientcmdapi.Cluster, error) {
-
+func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers, rootCAPubKeys []string) (*clientcmdapi.Cluster, error) {
 	tokenId, tokenSecret, err := tokenutil.ParseToken(discoveryToken)
 	if err != nil {
 		return nil, err
 	}
 
+	// Load the cfg.TLSDiscoveryRootCAPubKeys into a pubkeypin.Set
+	pubKeyPins := pubkeypin.NewSet()
+	err = pubKeyPins.Allow(rootCAPubKeys...)
+	if err != nil {
+		return nil, err
+	}
+
 	// The function below runs for every endpoint, and all endpoints races with each other.
 	// The endpoint that wins the race and completes the task first gets its kubeconfig returned below
 	baseKubeConfig := runForEndpointsAndReturnFirst(tokenAPIServers, func(endpoint string) (*clientcmdapi.Config, error) {
 
-		bootstrapConfig := buildInsecureBootstrapKubeConfig(endpoint)
-		clusterName := bootstrapConfig.Contexts[bootstrapConfig.CurrentContext].Cluster
-
-		client, err := kubeconfigutil.KubeConfigToClientSet(bootstrapConfig)
+		// clusterInfoURL is the URL of the bootstrap cluster info ConfigMap on the Kubernetes API
+		clusterInfoURL, err := url.Parse(fmt.Sprintf(
+			"https://%s/api/v1/namespaces/%s/configmaps/%s",
+			endpoint,
+			metav1.NamespacePublic,
+			bootstrapapi.ConfigMapClusterInfo))
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("invalid URL, can't connect: %v", err)
 		}
-
-		fmt.Printf("[discovery] Created cluster-info discovery client, requesting info from %q\n", bootstrapConfig.Clusters[clusterName].Server)
+		fmt.Printf("[discovery] Requesting cluster-info from %q\n", clusterInfoURL)
 
 		var clusterinfo *v1.ConfigMap
+		var bootstrapCertificateChain []*x509.Certificate
+
 		wait.PollImmediateInfinite(constants.DiscoveryRetryInterval, func() (bool, error) {
-			var err error
-			clusterinfo, err = client.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
+			response, certificates, err := withInsecureHTTPClient(func(client *http.Client) (*http.Response, error) {
+				return client.Get(clusterInfoURL.String())
+			})
 			if err != nil {
 				fmt.Printf("[discovery] Failed to request cluster info, will try again: [%s]\n", err)
 				return false, nil
 			}
+			defer response.Body.Close()
+
+			// Read the entire response body
+			responseBytes, err := ioutil.ReadAll(response.Body)
```

Comment:
You don't trust the server at this point. Probably good practice to limit the number of bytes read to something reasonable.